keras_nlp/models/gpt_neo_x/gpt_neo_x_decoder.py

```diff
@@ -0,0 +1,219 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.models.gpt_neo_x.gpt_neo_x_attention import GPTNeoXAttention
+from keras_nlp.utils.keras_utils import clone_initializer
+
+from keras_nlp.layers.transformer_layer_utils import (  # isort:skip
+    compute_causal_mask,
+    merge_padding_and_attention_mask,
+)
+
+
+class GPTNeoXDecoder(keras.layers.Layer):
+    """GPTNeoX decoder.
+
+    This class follows the architecture of the GPT-NeoX decoder layer in the
+    paper [GPT-NeoX-20B: An Open-Source Autoregressive Language Model](https://arxiv.org/abs/2204.06745). Users
+    can instantiate multiple instances of this class to stack up a decoder.
+
+    This layer will always apply a causal mask to the decoder attention layer.
+    This layer will correctly compute an attention mask from an implicit
+    Keras padding mask (for example, by passing `mask_zero=True` to a
+    `keras.layers.Embedding` layer). See the Masking and Padding
+    [guide](https://keras.io/guides/understanding_masking_and_padding/)
+    for more details.
+
+    This layer can be called with either one input which is as follows:
```

Comment:
just cut this paragraph, it looks like this is borrowed from TransformerDecoder but doesn't really apply here.